### PR TITLE
Express

### PR DESCRIPTION
**Pull request by seanoch**

### CODE REVIEW

File: 4-Express/server/mychecklist_server.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+const express = require('express')
```
Comment (dlvhdr) [inline 4-Express/server/mychecklist_server.js:1]: `my` is not a good name :) consider renaming to `todo_server.js`

File: 4-Express/client/scripts/communication_manager.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+const axios = require("axios").default;
```
Comment (dlvhdr) [inline 4-Express/client/scripts/communication_manager.js:1]: what a beast name, maybe be more descriptive: `todos_api.js`?

**Review by dlvhdr — CHANGES_REQUESTED** [review]

File: 4-Express/client/scripts/communication_manager.js (hunk 2 above)
Comment (tombenezra) [inline 4-Express/client/scripts/communication_manager.js:1]: why `.default`? use es6 imports for client code for consistency

File: 4-Express/client/scripts/communication_manager.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+const SERVER_URL = "http://localhost:3000";
+
+function addOrUpdateItem(id, item) {
+  const urlWithId = `${SERVER_URL}/${id}`;
+
+  axios.post(urlWithId, item).then(
```
Comment (tombenezra) [inline 4-Express/client/scripts/communication_manager.js:7]: return the result of `axios.post` from this function so whoever consumes this api will wait for the promise to be fullfilled and not the function itself.
e.g:
`return axios.post(...).catch(withErrorHandling)`
and the part which consumes it
`addOrUpdateItem(...).then(/*do something*/)`

File: 4-Express/client/scripts/communication_manager.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    (error) => {
+      console.log(error);
+    }
+  );
+}
+
+function removeItem(id) {
+  const urlWithId = `${SERVER_URL}/${id}`;
+
+  axios.delete(urlWithId, { params: { id: id } }).then(
```
Comment (tombenezra) [inline 4-Express/client/scripts/communication_manager.js:19]: same here

File: 4-Express/client/scripts/communication_manager.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+function addOrUpdateItem(id, item) {
+  const urlWithId = `${SERVER_URL}/${id}`;
+
+  axios.post(urlWithId, item).then(
+    (response) => {
+    },
+    (error) => {
+      console.log(error);
+    }
```
Comment (tombenezra) [inline 4-Express/client/scripts/communication_manager.js:12]: this amazing error handling technique is repeated several times. export to a function

File: 4-Express/client/scripts/communication_manager.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  );
+}
+
+async function getItems() {
+  try {
+    const response = await axios.get(SERVER_URL);
+    return response.data;
```
Comment (tombenezra) [inline 4-Express/client/scripts/communication_manager.js:31]: remember that `await` is blocking the execution so `line 31` will be delayed until the http request is resolved. 
consider returning the promise and let the consumer decide what to do with it

File: 4-Express/client/scripts/todo_list.js [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+    if (!valid) {
+      alert("Invalid item name!");
+    }
+
+    return valid;
+  }
+
+  function getItemElement(id, type) {
+    let element = document.querySelector("[data-id='" + id + "']");
```
Comment (tombenezra) [inline 4-Express/client/scripts/todo_list.js:114]: string interpolation can fit well here
```suggestion
    let element = document.querySelector(`[data-id='"${id}"']`);
```

File: 4-Express/client/scripts/todo_list.js [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+    addOrUpdateItem(id, {
+      text: label.innerText,
+      check: isChecked,
+    });
+  }
+
+  function removeFromServer(id) {
+    removeItem(id);
+  }
```
Comment (tombenezra) [inline 4-Express/client/scripts/todo_list.js:221]: redundant?

File: 4-Express/server/mychecklist_server.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    extended: true
+  })
+);
+app.use(express.json());
+
+app.get('/', (req, res) => {
+  res.sendJson(items);
+})
+
+app.post('/:id', (req, res) => {
```
Comment (tombenezra) [inline 4-Express/server/mychecklist_server.js:20]: The REST api convention for saving new entry should accepts no parameter from the url itself. so maybe have the id passed as part of the body as well?

File: 4-Express/server/mychecklist_server.js (hunk 1 above)
Comment (tombenezra) [inline 4-Express/server/mychecklist_server.js:1]: the name of the file is a bit weird. how about `todos_service` or something similar? 
you could also split this file into `index.js` that will configure and start the server and `todos_[controller/service].js` to manage the routes

**Review by tombenezra — CHANGES_REQUESTED** [review]

**Review by tombenezra — APPROVED** [review]

**Pull request merged**